src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/object/ObjectKeysFunctionIterator.java

```diff
@@ -118,15 +107,45 @@ public void setNextResult() {
     }
 
     @Override
-    public JavaRDD<Item> getRDDAux(DynamicContext dynamicContext) {
-        _currentDynamicContext = dynamicContext;
-        JavaRDD<Item> childRDD = _iterator.getRDD(dynamicContext);
-        FlatMapFunction<Item, Item> transformation = new ObjectKeysClosure();
-        return childRDD.flatMap(transformation).distinct();
+    public Item nextLocal() {
+        if (this._hasNext) {
+            Item result = _nextResults.remove();
+            if (_nextResults.isEmpty()) {
+                if (_iterator.isDataFrame()) {
+                    this._hasNext = false;
+                } else {
+                    setResultsFromNextObjectItem();
+                }
+            }
+            return result;
+        }
+        throw new IteratorFlowException(
+                RuntimeIterator.FLOW_EXCEPTION_MESSAGE + " KEYS function",
+                getMetadata()
+        );
     }
 
     @Override
-    public boolean initIsRDD() {
-        return _iterator.isRDD();
+    protected boolean hasNextLocal() {
+        return _hasNext;
+    }
+
+    @Override
+    protected void resetLocal(DynamicContext context) {
+        _iterator.reset(_currentDynamicContext);
+        setResultsFromNextObjectItem();
```

Comment:
Here the same switch should be made as in openLocal(), as it could be a DF.

Reply:
Fixed now